web-app/js/smartR/heatmapService.js

```diff
@@ -5,44 +5,104 @@
  */
 
 HeatmapService = (function(smartRHeatmap){
+    var CHECK_DELAY = 1000
+    var NOOP_ABORT = function() {};
 
     var service = {
-        statusInterval : 0
+        currentRequestAbort: NOOP_ABORT,
+        lastFetchedLabels: [],
     };
 
-    /**
-     *
-     * @param files
-     * @returns {{}}
-     * @private
-     */
-    var _getSummaryFiles = function (files) {
-        var retv = {}, types = ['png', 'json'];
-
-        types.forEach(function (type) {
-            retv[type] = files.filter(function (file) {
-                console.log('type',type);
-                console.log('file',file);
-                console.log('file.indexOf(type)',file.indexOf(type));
-                return file.indexOf(type) < 0  ? null : file;
+    var _setStatusRequestTimeout = function() {
+        var timeout = setTimeout.apply(undefined, arguments)
+        service.currentRequestAbort = function() {
+            clearTimeout(timeout);
+            service.currentRequestAbort = NOOP_ABORT;
+        }
+    }
+
+    /* generate unique labels for the concept paths. It recursively resolves
+     * clashes. */
+    var _generateLabels = (function() {
```

Comment:
as discussed, this way would have impact on R esp when it uses the labels to create result files